onadata/apps/logger/data_migration/surveytree.py

```diff
@@ -1,68 +1,100 @@
+from functools import partial
+from itertools import ifilter
 from lxml import etree
 
+from .xmltree import XMLTree
+from .common import compose, concat_map
+
 
 class MissingFieldException(Exception):
     pass
 
 
-class SurveyTree(object):
+class SurveyTree(XMLTree):
     """
     Parse XForm Instance from xml string into tree.
     """
-    # XML elements that should not be considered as fields
-    NOT_RELEVANT = ['formhub', 'meta', 'imei']
-
     def __init__(self, survey):
         # Handle both cases when instance or string is passed.
         try:
             self.root = etree.XML(survey.xml)
         except AttributeError:
             self.root = etree.XML(survey)
 
-    def __repr__(self):
-        return self.to_string()
+    def get_fields(self):
+        """Parse and return list of all fields in form."""
+        return self.retrieve_leaf_elems(self.root)
 
-    def to_string(self, pretty=True):
-        return etree.tostring(self.root, pretty_print=pretty)
+    def get_groups(self):
+        return concat_map(self.retrieve_groups, self.root.getchildren())
 
-    def get_fields(self):
-        """Return fields as list with tree Elements."""
-        return [
-            field for field in self.root.getchildren()
-            if field.tag not in self.NOT_RELEVANT
-        ]
+    def get_all_elems(self):
+        """Return a list of both groups and fields"""
+        return concat_map(self.retrieve_all_elems, self.root.getchildren())
 
     def get_fields_names(self):
         """Return fields as list of string with field names."""
-        return [
-            field.tag for field in self.root.getchildren()
-            if field.tag not in self.NOT_RELEVANT
-        ]
+        return map(lambda f: f.tag, self.get_fields())
+
+    def get_groups_names(self):
+        """Return fields as list of string with field names."""
+        return map(lambda g: g.tag, self.get_groups())
+
+    def _get_matching_elems(self, condition_func):
+        """Return elems that match condition"""
+        return iter(filter(condition_func, self.get_all_elems()))
```

Comment:
if you want iterator for python2, use list comprehension notation here (condition_func(elem) for elem in self.get_all_elems()), since filter will create list

Reply:
thanks, used `ifilter` instead